### PR TITLE
Explain how `version` works for `git` dependencies

### PR DESCRIPTION
**Pull request by BatmanAoD**
### What does this PR try to resolve?

Explains how `version` works for `git` dependencies. This confused me, and I couldn't find any mention of it e.g. on the user forums.

### How should we test and review this PR?

Documentation-only change

### CODE REVIEW

**Review by weihanglo — COMMENTED** [review]
Thanks for the PR. Just wonder if [Multiple locations](https://doc.rust-lang.org/nightly/cargo/reference/specifying-dependencies.html#multiple-locations) section already cover this, no?

**Comment by BatmanAoD** [comment]
@weihanglo It actually does make things a lot clearer; I'll take another look and revise my suggestion. I didn't actually look at that section before because I didn't realize that the `version` key would still cause an attempt to look up the crate in a registry, so I didn't think that "multiple locations" even applied to this situation.

**Comment by BatmanAoD** [comment]
@weihanglo Sorry for not getting back to this sooner. How does it look now?

**Review by weihanglo — COMMENTED** [review]
`path` dependencies also have the exact behavior. Is it possible to merge this into the notice below, and update for both `path` and `git` dependencies to share the similar text?

https://github.com/rust-lang/cargo/blob/5b377cece0e0dd0af686cf53ce4637d5d85c2a10/src/doc/src/reference/specifying-dependencies.md?plain=1#L168-L170

So maybe something like:

```markdown
> **Note**: A dependency can have both a `git` and `version` keys, these are
> considered separate locations for resolving the dependency, as [crates.io]
> does not allow packages to be published with `git` dependencies (`git`
> [dev-dependencies] are ignored). See the [Multiple locations](#multiple-locations)
> section for a fallback alternative.
```

What do you think?

(I tried to make it concise but I believe you could do this better that me)

**Comment by weihanglo** [comment]
let me take over the review

r? @weihanglo

**Comment by BatmanAoD** [comment]
I think the note about `crates.io` not permitting packages with `git` dependencies is saying something fairly different, namely that users are not permitted to publish packages on `crates.io` that import code from somewhere other than `crates.io` itself. (This note doesn't explicitly mention third-party registries, but the one a couple sections above does.)

I've made some more extensive changes to reflect that; please let me know what you think.

(The introduction of the "requirements syntax" header, and demoting of the different version-requirement syntax headers to go under that header, is technically unrelated, so let me know if I need to undo that change.)

File: src/doc/src/reference/specifying-dependencies.md [hunk 1]
```diff
@@ -195,16 +197,27 @@ varies by where the repo is hosted; GitHub in particular exposes a reference to
 the most recent commit of every pull request as shown, but other git hosts often
 provide something equivalent, possibly under a different naming scheme.
 
+> **Note**: Neither the `git` key nor the `path` key changes the meaning of the
```
Comment (weihanglo) [inline src/doc/src/reference/specifying-dependencies.md:200]: Looks pretty good! Thank you for the perfect explanation.

Just wonder why insert this in the middle of the section, but not at the end?

Comment (BatmanAoD) [reply src/doc/src/reference/specifying-dependencies.md:200]: I guess I didn't want two notes in a row, and this seemed like an okay place, but I ended up putting two notes in a row on the path section anyway, so it probably makes sense to do that here too.

Comment (weihanglo) [reply src/doc/src/reference/specifying-dependencies.md:200]: Sounds good to me. Let's put it at the bottom.

Comment (BatmanAoD) [reply src/doc/src/reference/specifying-dependencies.md:207]: Moved!

**Review by weihanglo — APPROVED** [review]
Great work! Thanks!

**Comment by weihanglo** [comment]
Hmm… I guess bors was stuck, so it was never merged successfully. @BatmanAoD, could you reopen so that we can merge it again?

**Comment by BatmanAoD** [comment]
@weihanglo Oops, for some reason I thought it was safe to delete the branch. Restored and re-opened.

**Comment by weihanglo** [comment]
No worries!

@bors r+

**Pull request merged**